staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1864,3 +1986,523 @@ func TestForgetWatcher(t *testing.T) {
+		opts.Predicate = p
+		opts.SendInitialEvents = sendInitialEvents
+		opts.ResourceVersion = rv
+		return opts
+	}
+
+	scenarios := []struct {
+		name                      string
+		opts                      storage.ListOptions
+		requiredResourceVersion   int
+		watchCacheResourceVersion int
+
+		expectedBookmarkResourceVersion int
+	}{
+		// +-----------------+---------------------+-----------------------+
+		// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+		// +=================+=====================+=======================+
+		// | Unset           | true/false          | nil/true/false        |
+		// +-----------------+---------------------+-----------------------+
+		{
+			name:                      "RV=unset, allowWatchBookmarks=true, sendInitialEvents=nil, expectedBookmarkRV=0",
+			opts:                      listOptions(true, nil, ""),
+			requiredResourceVersion:   100,
+			watchCacheResourceVersion: 90,
+		},
+		{
+			// TODO: this looks incorrect
+			name:                            "RV=unset, allowWatchBookmarks=true, sendInitialEvents=true, expectedBookmarkRV=90",
+			opts:                            listOptions(true, pointer.Bool(true), ""),
+			requiredResourceVersion:         100,
+			watchCacheResourceVersion:       90,
+			expectedBookmarkResourceVersion: 90,
+		},
+		{
+			name:                      "RV=unset, allowWatchBookmarks=true, sendInitialEvents=false, expectedBookmarkRV=0",
+			opts:                      listOptions(true, pointer.Bool(false), ""),
+			requiredResourceVersion:   100,
+			watchCacheResourceVersion: 90,
+		},
+		{
+			name:                      "RV=unset, allowWatchBookmarks=false, sendInitialEvents=nil, expectedBookmarkRV=0",
+			opts:                      listOptions(false, nil, ""),
+			requiredResourceVersion:   100,
+			watchCacheResourceVersion: 90,
+		},
+		{
+			name:                      "RV=unset, allowWatchBookmarks=false, sendInitialEvents=true, expectedBookmarkRV=0",
+			opts:                      listOptions(false, pointer.Bool(true), ""),
+			requiredResourceVersion:   100,
+			watchCacheResourceVersion: 90,
+		},
+		{
+			name:                      "RV=unset, allowWatchBookmarks=false, sendInitialEvents=false, expectedBookmarkRV=0",
+			opts:                      listOptions(false, pointer.Bool(false), ""),
+			requiredResourceVersion:   100,
+			watchCacheResourceVersion: 90,
+		},
+		// +-----------------+---------------------+-----------------------+
+		// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+		// +=================+=====================+=======================+
+		// | 0               | true/false          | nil/true/false        |
+		// +-----------------+---------------------+-----------------------+
+		{
+			name:                      "RV=0, allowWatchBookmarks=true, sendInitialEvents=nil, expectedBookmarkRV=0",
+			opts:                      listOptions(true, nil, "0"),
+			requiredResourceVersion:   100,
```

Comment:
I don't think required would be set to 100 in those cases - it would be 0, no?

Reply:
If we set it to `0` then we cannot distinguish between `RV=0` and `requiredResourceVersion=0` by setting it to some artificial RV we can assert it won't be returned.

Reply:
I don't care if assert that something will/won't be returned - I want to test what exactly will be passed in production and check what our function returns in such case.

Reply:
This is a simplified test that tests only a single function, if we would like to test what will be passed in production then i think we would have to develop a different test.

If there was a bug in that function that would return `requiredResourceVersion` for this case, then we wouldn't have caught it because it is exactly the same as `RV=0`, right ?